File: src/main/java/io/github/moulberry/notenoughupdates/profileviewer/bestiary/BestiaryPage.java
/*
 * Copyright (C) 2022 NotEnoughUpdates contributors
 *
 * This file is part of NotEnoughUpdates.
 *
 * NotEnoughUpdates is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * NotEnoughUpdates is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with NotEnoughUpdates. If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.moulberry.notenoughupdates.profileviewer.bestiary;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import io.github.moulberry.notenoughupdates.core.util.StringUtils;
import io.github.moulberry.notenoughupdates.profileviewer.GuiProfileViewer;
import io.github.moulberry.notenoughupdates.profileviewer.GuiProfileViewerPage;
import io.github.moulberry.notenoughupdates.profileviewer.ProfileViewer;
import io.github.moulberry.notenoughupdates.profileviewer.ProfileViewerUtils;
import io.github.moulberry.notenoughupdates.profileviewer.SkyblockProfiles;
import io.github.moulberry.notenoughupdates.util.Constants;
import io.github.moulberry.notenoughupdates.util.Utils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.opengl.GL11;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

import static io.github.moulberry.notenoughupdates.profileviewer.GuiProfileViewer.pv_elements;

public class BestiaryPage extends GuiProfileViewerPage {

	private static final ResourceLocation BESTIARY_TEXTURE = new ResourceLocation("notenoughupdates:pv_bestiary_tab.png");
	private static final int XCOUNT = 7;
	private static final int YCOUNT = 5;
	private static final float XPADDING = (190 - XCOUNT * 20) / (float) (XCOUNT + 1);
	private static final float YPADDING = (202 - YCOUNT * 20) / (float) (YCOUNT + 1);
	private ItemStack selectedBestiaryLocation = null;
	private List<String> tooltipToDisplay = null;

	public BestiaryPage(GuiProfileViewer instance) {
		super(instance);
	}

	@Override
	public void drawPage(int mouseX, int mouseY, float partialTicks) {
		int guiLeft = GuiProfileViewer.getGuiLeft();
		int guiTop = GuiProfileViewer.getGuiTop();

		SkyblockProfiles.SkyblockProfile selectedProfile = getSelectedProfile();
		if (selectedProfile == null) {
			return;
		}

		JsonObject profileInfo = selectedProfile.getProfileJson();

		int bestiarySize = BestiaryData.getBestiaryLocations().size();
		int bestiaryXSize = (int) (350f / (bestiarySize - 1 + 0.0000001f));

		{
			int yIndex = 0;
			for (ItemStack stack : BestiaryData.getBestiaryLocations().keySet()) {
				Minecraft.getMinecraft().getTextureManager().bindTexture(pv_elements);
				if (mouseX > guiLeft + 30 + bestiaryXSize * yIndex && mouseX < guiLeft + 30 + bestiaryXSize * yIndex + 20) {
					if (mouseY > guiTop + 10 && mouseY < guiTop + 10 + 20) {
						tooltipToDisplay = stack.getTooltip(Minecraft.getMinecraft().thePlayer, false);
					}
				}
				if (stack == selectedBestiaryLocation) {
					Utils.drawTexturedRect(
						guiLeft + 30 + bestiaryXSize * yIndex,
						guiTop + 10,
						20,
						20,
						20 / 256f,
						0,
						20 / 256f,
						0,
						GL11.GL_NEAREST
					);
					Utils.drawItemStack(stack, guiLeft + 32 + bestiaryXSize * yIndex, guiTop + 12);
				} else {
					Utils.drawTexturedRect(
						guiLeft + 30 + bestiaryXSize * yIndex,
						guiTop + 10,
						20,
						20,
						0,
						20 / 256f,
						0,
						20 / 256f,
						GL11.GL_NEAREST
					);
					Utils.drawItemStack(stack, guiLeft + 32 + bestiaryXSize * yIndex, guiTop + 12);
				}
				yIndex++;
			}
		}
		ScaledResolution scaledResolution = new ScaledResolution(Minecraft.getMinecraft());
		int width = scaledResolution.getScaledWidth();
		int height = scaledResolution.getScaledHeight();

		Minecraft.getMinecraft().getTextureManager().bindTexture(BESTIARY_TEXTURE);
		Utils.drawTexturedRect(guiLeft, guiTop, 431, 202, GL11.GL_NEAREST);

		GlStateManager.color(1, 1, 1, 1);
		Color color = new Color(128, 128, 128, 255);
		Utils.renderAlignedString(
			EnumChatFormatting.RED + "Bestiary Level: ",
			EnumChatFormatting.GRAY + "" + (float) getSelectedProfile().getBestiaryLevel() / 10,
			guiLeft + 220,
			guiTop + 50,
			110
		);

		GlStateManager.disableLighting();
		RenderHelper.enableGUIStandardItemLighting();
		List<String> mobs = BestiaryData.getBestiaryLocations().get(selectedBestiaryLocation);
		if (mobs != null) {
			for (int i = 0; i < mobs.size(); i++) {
				String mob = mobs.get(i);
				if (mob != null) {
					ItemStack mobItem = BestiaryData.getBestiaryMobs().get(mob);
					if (mobItem != null) {
						int xIndex = i % XCOUNT;
						int yIndex = i / XCOUNT;

						float x = 23 + XPADDING + (XPADDING + 20) * xIndex;
						float y = 30 + YPADDING + (YPADDING + 20) * yIndex;

						float completedness = 0;

						GlStateManager.color(1, 1, 1, 1);
						Minecraft.getMinecraft().getTextureManager().bindTexture(pv_elements);
						Utils.drawTexturedRect(
							guiLeft + x,
							guiTop + y,
							20,
							20 * (1 - completedness),
							0,
							20 / 256f,
							0,
							20 * (1 - completedness) / 256f,
							GL11.GL_NEAREST
						);
						//GlStateManager.color(1, 185 / 255f, 0, 1);
						Minecraft.getMinecraft().getTextureManager().bindTexture(pv_elements);
						Utils.drawTexturedRect(
							guiLeft + x,
							guiTop + y + 20 * (1 - completedness),
							20,
							20 * (completedness),
							0,
							20 / 256f,
							20 * (1 - completedness) / 256f,
							20 / 256f,
							GL11.GL_NEAREST
						);
						Utils.drawItemStack(mobItem, guiLeft + (int) x + 2, guiTop + (int) y + 2);
						float kills = Utils.getElementAsFloat(Utils.getElement(profileInfo, "bestiary.kills_" + mob), 0);
						float deaths = Utils.getElementAsFloat(Utils.getElement(profileInfo, "bestiary.deaths_" + mob), 0);

						String type;
						if (BestiaryData.getMobType().get(mob) != null) {
							type = BestiaryData.getMobType().get(mob);
						} else {
							type = "MOB";
						}
						JsonObject leveling = Constants.LEVELING;
						ProfileViewer.Level level = null;
						if (leveling != null && Utils.getElement(leveling, "bestiary." + type) != null) {
							JsonArray levelingArray = Utils.getElement(leveling, "bestiary." + type).getAsJsonArray();
							int levelCap = Utils.getElementAsInt(Utils.getElement(leveling, "bestiary.caps." + type), 0);
							level = ProfileViewerUtils.getLevel(levelingArray, kills, levelCap, false);
						} else {
							Utils.showOutdatedRepoNotification();
						}

						float levelNum = -1;
						if (level != null) {
							levelNum = level.level;
						}
						if (mouseX > guiLeft + (int) x + 2 && mouseX < guiLeft + (int) x + 18) {
							if (mouseY > guiTop + (int) y + 2 && mouseY < guiTop + (int) y + 18) {
								tooltipToDisplay = new ArrayList<>();
								tooltipToDisplay.add(
									mobItem.getDisplayName() + " " + ((levelNum == -1) ? "?" : (int) Math.floor(levelNum))
								);
								tooltipToDisplay.add(
									EnumChatFormatting.GRAY + "Kills: " + EnumChatFormatting.GREEN + StringUtils.formatNumber(kills)
								);
								tooltipToDisplay.add(
									EnumChatFormatting.GRAY + "Deaths: " + EnumChatFormatting.GREEN + StringUtils.formatNumber(deaths)
								);
								if (level != null) {
									String progressStr;
									if (level.maxed) {
										progressStr = EnumChatFormatting.GOLD + "MAXED!";
									} else {
										progressStr = EnumChatFormatting.AQUA +
											StringUtils.shortNumberFormat(Math.round((levelNum % 1) * level.maxXpForLevel)) +
											"/" +
											StringUtils.shortNumberFormat(level.maxXpForLevel);
									}
									tooltipToDisplay.add(EnumChatFormatting.GRAY + "Progress: " + progressStr);
								}
							}
						}

						GlStateManager.color(1, 1, 1, 1);
						//						if (tier >= 0) {
						//							Utils.drawStringCentered(tierString,
						//								guiLeft + x + 10, guiTop + y - 4, true,
						//								tierStringColour
						//							);
						//						}
						Utils.drawStringCentered(
							(int) Math.floor(levelNum) + "",
							guiLeft + x + 10, guiTop + y + 26, true, color.getRGB()
						);
					}
				}
			}
		}
		if (tooltipToDisplay != null) {
			List<String> grayTooltip = new ArrayList<>(tooltipToDisplay.size());
			for (String line : tooltipToDisplay) {
				grayTooltip.add(EnumChatFormatting.GRAY + line);
			}
			Utils.drawHoveringText(grayTooltip, mouseX, mouseY, width, height, -1);
			tooltipToDisplay = null;
		}
	}

	@Override
	public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
		int guiLeft = GuiProfileViewer.getGuiLeft();
		int guiTop = GuiProfileViewer.getGuiTop();

		int bestiarySize = BestiaryData.getBestiaryLocations().size();
		int bestiaryYSize = (int) (350f / (bestiarySize - 1 + 0.0000001f));
		int yIndex = 0;
		for (ItemStack stack : BestiaryData.getBestiaryLocations().keySet()) {
			if (mouseX > guiLeft + 30 + bestiaryYSize * yIndex && mouseX < guiLeft + 30 + bestiaryYSize * yIndex + 20) {
				if (mouseY > guiTop + 10 && mouseY < guiTop + 10 + 20) {
					selectedBestiaryLocation = stack;
					Utils.playPressSound();
					return;
				}
			}
			yIndex++;
		}
	}
}

File: src/main/kotlin/io/github/moulberry/notenoughupdates/miscfeatures/profileviewer/bestiary/BestiaryPage.kt
/*
 * Copyright (C) 2022-2023 NotEnoughUpdates contributors
 *
 * This file is part of NotEnoughUpdates.
 *
 * NotEnoughUpdates is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * NotEnoughUpdates is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with NotEnoughUpdates. If not, see <https://www.gnu.org/licenses/>.
 */
package io.github.moulberry.notenoughupdates.miscfeatures.profileviewer.bestiary

import io.github.moulberry.notenoughupdates.core.util.StringUtils
import io.github.moulberry.notenoughupdates.profileviewer.GuiProfileViewer
import io.github.moulberry.notenoughupdates.profileviewer.GuiProfileViewerPage
import io.github.moulberry.notenoughupdates.miscfeatures.profileviewer.bestiary.BestiaryData.calculateTotalBestiaryTiers
import io.github.moulberry.notenoughupdates.miscfeatures.profileviewer.bestiary.BestiaryData.hasMigrated
import io.github.moulberry.notenoughupdates.miscfeatures.profileviewer.bestiary.BestiaryData.parseBestiaryData
import io.github.moulberry.notenoughupdates.util.Constants
import io.github.moulberry.notenoughupdates.util.Utils
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumChatFormatting
import net.minecraft.util.ResourceLocation
import org.lwjgl.input.Mouse
import org.lwjgl.opengl.GL11
import java.awt.Color

/**
 * Individual mob entry in the Bestiary
 */
data class Mob(
    val name: String, val icon: ItemStack, val kills: Double, val deaths: Double, val mobLevelData: MobLevelData
)

/**
 * A Bestiary category as defined in `constants/bestiary.json`
 */
data class Category(
    val id: String,
    val name: String,
    val icon: ItemStack,
    val mobs: List<Mob>,
    val subCategories: List<Category>,
    val familyData: FamilyData
)

/**
 * Level data for one specific mob
 */
data class MobLevelData(
    val level: Int, val maxLevel: Boolean, val progress: Double, val totalProgress: Double, val mobKillData: MobKillData
)

/**
 * Kills data for one specific mob
 */
data class MobKillData(
    val tierKills: Double, val tierReq: Double, val cappedKills: Double, val cap: Double
)

/**
 * Family data for one specific category
 */
data class FamilyData(
    val found: Int, val completed: Int, val total: Int
)

class BestiaryPage(instance: GuiProfileViewer?) : GuiProfileViewerPage(instance) {
    private var selectedCategory = "dynamic"
    private var selectedSubCategory = ""
    private var tooltipToDisplay: MutableList<String> = mutableListOf()
    private var bestiaryLevel = 0.0
    private var computedCategories: MutableList<Category> = mutableListOf()
    private var lastProfileName = ""

    private val bestiaryTexture = ResourceLocation("notenoughupdates:pv_bestiary_tab.png")
    private val mobListXCount = 9
    private val mobListYCount = 5
    private val mobListXPadding = (240 - mobListXCount * 20) / (mobListXCount + 1).toFloat()
    private val mobListYPadding = (202 - mobListYCount * 20) / (mobListYCount + 1).toFloat()

    override fun drawPage(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val guiLeft = GuiProfileViewer.getGuiLeft()
        val guiTop = GuiProfileViewer.getGuiTop()

        val selectedProfile = GuiProfileViewer.getSelectedProfile() ?: return
        val profileInfo = selectedProfile.profileJson
        val profileName = GuiProfileViewer.getProfileName()

        if (!hasMigrated(profileInfo) || Constants.BESTIARY == null) {
            Utils.drawStringCentered(
                "${EnumChatFormatting.RED}No valid bestiary data!",
                guiLeft + 431 / 2f,
                (guiTop + 101).toFloat(),
                true,
                0
            )
            lastProfileName = profileName
            return
        }
        // Do the initial parsing only once or on profile switch
        if (computedCategories.isEmpty() || lastProfileName != profileName) {
            computedCategories = parseBestiaryData(profileInfo)
            bestiaryLevel = calculateTotalBestiaryTiers(computedCategories).toDouble()
        }

        if (computedCategories.isEmpty() || Constants.BESTIARY == null) {
            Utils.drawStringCentered(
                "${EnumChatFormatting.RED}No valid bestiary data!",
                guiLeft + 431 / 2f,
                (guiTop + 101).toFloat(),
                true,
                0
            )
            lastProfileName = profileName
            return
        }
        lastProfileName = profileName

        val bestiarySize = computedCategories.size
        val bestiaryXSize = (350f / (bestiarySize - 1 + 0.0000001f)).toInt()


        // Render the category list
        for ((categoryXIndex, category) in computedCategories.withIndex()) {
            Minecraft.getMinecraft().textureManager.bindTexture(GuiProfileViewer.pv_elements)

            if (mouseX > guiLeft + 30 + bestiaryXSize * categoryXIndex &&
                mouseX < guiLeft + 30 + bestiaryXSize * categoryXIndex + 20
                && mouseY > guiTop + 10 && mouseY < guiTop + 10 + 20
            ) {
                tooltipToDisplay.add(EnumChatFormatting.GRAY.toString() + category.name)
                if (Mouse.getEventButtonState() && selectedCategory != category.id) {
                    selectedCategory = category.id
                    Utils.playPressSound()
                }
            }


            if (category.id == selectedCategory) {
                Utils.drawTexturedRect(
                    (guiLeft + 30 + bestiaryXSize * categoryXIndex).toFloat(),
                    (guiTop + 10).toFloat(),
                    20f,
                    20f,
                    20 / 256f,
                    0f,
                    20 / 256f,
                    0f,
                    GL11.GL_NEAREST
                )
            } else {
                Utils.drawTexturedRect(
                    (guiLeft + 30 + bestiaryXSize * categoryXIndex).toFloat(),
                    (guiTop + 10).toFloat(),
                    20f,
                    20f,
                    0f,
                    20 / 256f,
                    0f,
                    20 / 256f,
                    GL11.GL_NEAREST
                )
            }
            Utils.drawItemStack(category.icon, guiLeft + 32 + bestiaryXSize * categoryXIndex, guiTop + 12)
        }

        val scaledResolution = ScaledResolution(Minecraft.getMinecraft())
        val width = scaledResolution.scaledWidth
        val height = scaledResolution.scaledHeight

        Minecraft.getMinecraft().textureManager.bindTexture(bestiaryTexture)
        Utils.drawTexturedRect(guiLeft.toFloat(), guiTop.toFloat(), 431f, 202f, GL11.GL_NEAREST)
        GlStateManager.color(1f, 1f, 1f, 1f)
        val color = Color(128, 128, 128, 255)
        Utils.renderAlignedString(
            EnumChatFormatting.RED.toString() + "Milestone: ",
            "${EnumChatFormatting.GRAY}${(bestiaryLevel / 10)}",
            (guiLeft + 280).toFloat(),
            (guiTop + 50).toFloat(),
            110
        )

        // Render the subcategories in the bottom right corner, if applicable
        val selectedCategory = computedCategories.first { it.id == selectedCategory }
        if (selectedCategory.subCategories.isNotEmpty()) {
            if (selectedSubCategory == "") {
                selectedSubCategory = selectedCategory.subCategories.first().id
            }

            Utils.renderShadowedString(
                "${EnumChatFormatting.RED}Subcategories", (guiLeft + 317).toFloat(), (guiTop + 165).toFloat(), 1000
            )
            GlStateManager.color(1f, 1f, 1f, 1f)

            val xStart = (guiLeft + 280).toFloat()
            val y = (guiTop + 175).toFloat()
            for ((i, subCategory) in selectedCategory.subCategories.withIndex()) {
                Minecraft.getMinecraft().textureManager.bindTexture(GuiProfileViewer.pv_elements)

                if (subCategory.id == selectedSubCategory) {
                    Utils.drawTexturedRect(
                        (xStart + 24 * i.toFloat()),
                        y,
                        20f,
                        20f,
                        20 / 256f,
                        0f,
                        20 / 256f,
                        0f,
                        GL11.GL_NEAREST
                    )
                } else {
                    Utils.drawTexturedRect(
                        (xStart + 24 * i.toFloat()),
                        y,
                        20f,
                        20f,
                        0f,
                        20 / 256f,
                        0f,
                        20 / 256f,
                        GL11.GL_NEAREST
                    )
                }
                Utils.drawItemStack(subCategory.icon, (xStart + 24 * i + 2).toInt(), y.toInt() + 2)
                if (mouseX > xStart + 24 * i
                    && mouseX < xStart + 24 * (i + 1)
                    && mouseY > y
                    && mouseY < y + 16
                ) {
                    tooltipToDisplay.add(subCategory.name)
                    if (Mouse.getEventButtonState() && selectedSubCategory != subCategory.id) {
                        selectedSubCategory = subCategory.id
                        Utils.playPressSound()
                    }
                }
            }
        } else {
            selectedSubCategory = ""
        }

        // Render family information
        var catData = selectedCategory.familyData
        Utils.renderAlignedString(
            EnumChatFormatting.RED.toString() + "Families Found:",
            (if (catData.found == catData.total) "§6" else "§7") + "${catData.found}/${catData.total}",
            guiLeft + 280F,
            guiTop + 70F,
            110
        )
        if (catData.found == catData.total) {
            instance.renderGoldBar(guiLeft + 280F, guiTop + 80F, 112F)
        } else {
            instance.renderBar(guiLeft + 280F, guiTop + 80F, 112F, catData.found / catData.total.toFloat())
        }

        Utils.renderAlignedString(
            EnumChatFormatting.RED.toString() + "Families Completed:",
            (if (catData.completed == catData.total) "§6" else "§7") + "${catData.completed}/${catData.total}",
            guiLeft + 280F,
            guiTop + 90F,
            110
        )
        if (catData.completed == catData.total) {
            instance.renderGoldBar(guiLeft + 280F, guiTop + 100F, 112F)
        } else {
            instance.renderBar(guiLeft + 280F, guiTop + 100F, 112F, catData.completed / catData.total.toFloat())
        }

        // Render subcategory family information, if possible
        if (selectedSubCategory != "") {
            catData = selectedCategory.subCategories.find { it.id == selectedSubCategory }!!.familyData
            Utils.renderAlignedString(
                EnumChatFormatting.RED.toString() + "Families Found:",
                (if (catData.found == catData.total) "§6" else "§7") + "${catData.found}/${catData.total}",
                guiLeft + 280F,
                guiTop + 120F,
                110
            )
            if (catData.found == catData.total) {
                instance.renderGoldBar(guiLeft + 280F, guiTop + 130F, 112F)
            } else {
                instance.renderBar(guiLeft + 280F, guiTop + 130F, 112F, catData.found / catData.total.toFloat())
            }

            Utils.renderAlignedString(
                EnumChatFormatting.RED.toString() + "Families Completed:",
                (if (catData.completed == catData.total) "§6" else "§7") + "${catData.completed}/${catData.total}",
                guiLeft + 280F,
                guiTop + 140F,
                110
            )
            if (catData.completed == catData.total) {
                instance.renderGoldBar(guiLeft + 280F, guiTop + 150F, 112F)
            } else {
                instance.renderBar(guiLeft + 280F, guiTop + 150F, 112F, catData.completed / catData.total.toFloat())
            }
        }

        // Determine which mobs should be displayed
        val mobs = if (selectedSubCategory != "") {
            selectedCategory.subCategories.first { it.id == selectedSubCategory }.mobs
        } else {
            selectedCategory.mobs
        }

        // Render the mob list
        for ((i, mob) in mobs.withIndex()) {
            val stack = mob.icon
            val xIndex = i % mobListXCount
            val yIndex = i / mobListXCount
            val x = 23 + mobListXPadding + (mobListXPadding + 20) * xIndex
            val y = 30 + mobListYPadding + (mobListYPadding + 20) * yIndex

            GlStateManager.disableLighting()
            RenderHelper.enableGUIStandardItemLighting()
            GlStateManager.color(1f, 1f, 1f, 1f)
            Minecraft.getMinecraft().textureManager.bindTexture(GuiProfileViewer.pv_elements)
            Utils.drawTexturedRect(
                guiLeft + x,
                guiTop + y,
                20f,
                20f,
                0f,
                20 / 256f,
                0f,
                20f / 256f,
                GL11.GL_NEAREST
            )
            Utils.drawItemStack(stack, guiLeft + x.toInt() + 2, guiTop + y.toInt() + 2)
            val kills = mob.kills
            val deaths = mob.deaths

            if (mouseX > guiLeft + x.toInt() + 2 && mouseX < guiLeft + x.toInt() + 18) {
                if (mouseY > guiTop + y.toInt() + 2 && mouseY < guiTop + y.toInt() + 18) {
                    tooltipToDisplay = ArrayList()
                    tooltipToDisplay.add(
                        "${mob.name} ${mob.mobLevelData.level}"
                    )
                    tooltipToDisplay.add(
                        EnumChatFormatting.GRAY.toString() + "Kills: " + EnumChatFormatting.GREEN +
                                StringUtils.formatNumber(kills)
                    )
                    tooltipToDisplay.add(
                        EnumChatFormatting.GRAY.toString() + "Deaths: " + EnumChatFormatting.GREEN +
                                StringUtils.formatNumber(deaths)
                    )
                    tooltipToDisplay.add("")

                    if (!mob.mobLevelData.maxLevel) {
                        tooltipToDisplay.add(
                            EnumChatFormatting.GRAY.toString() + "Progress to Tier ${mob.mobLevelData.level + 1}: " +
                                    EnumChatFormatting.AQUA + "${mob.mobLevelData.progress}%"
                        )

                        var bar = "§3§l§m"
                        for (j in 1..14) {
                            var col = ""
                            if (mob.mobLevelData.progress < j * (100 / 14)) col = "§f§l§m"
                            bar += "$col "
                        }
                        tooltipToDisplay.add(
                            "${bar}§r§b ${StringUtils.formatNumber(mob.mobLevelData.mobKillData.tierKills)}/${
                                StringUtils.formatNumber(
                                    mob.mobLevelData.mobKillData.tierReq
                                )
                            }"
                        )
                        tooltipToDisplay.add("")
                    }

                    tooltipToDisplay.add(
                        EnumChatFormatting.GRAY.toString() + "Overall Progress: " + EnumChatFormatting.AQUA + "${mob.mobLevelData.totalProgress}%" +
                                if (mob.mobLevelData.maxLevel) " §7(§c§lMAX!§r§7)" else ""
                    )
                    var bar = "§3§l§m"
                    for (j in 1..14) {
                        var col = ""
                        if (mob.mobLevelData.totalProgress < j * (100 / 14)) col = "§f§l§m"
                        bar += "$col "
                    }
                    tooltipToDisplay.add(
                        "${bar}§r§b ${StringUtils.formatNumber(mob.mobLevelData.mobKillData.cappedKills)}/${
                            StringUtils.formatNumber(
                                mob.mobLevelData.mobKillData.cap
                            )
                        }"
                    )
                }
            }
            GlStateManager.color(1f, 1f, 1f, 1f)
            Utils.drawStringCentered(
                if (mob.mobLevelData.maxLevel) {
                    "${EnumChatFormatting.GOLD}${mob.mobLevelData.level}"
                } else {
                    mob.mobLevelData.level.toString()
                }, guiLeft + x + 10, guiTop + y + 26, true, color.rgb
            )
        }

        // Render the accumulated tooltip, if applicable
        if (tooltipToDisplay.isNotEmpty()) {
            val grayTooltip: MutableList<String> = ArrayList(tooltipToDisplay.size)
            for (line in tooltipToDisplay) {
                grayTooltip.add(EnumChatFormatting.GRAY.toString() + line)
            }
            Utils.drawHoveringText(grayTooltip, mouseX, mouseY, width, height, -1)
            tooltipToDisplay.clear()
        }
    }
}
